### ISSUE TITLE
A note on the implementation.

### ISSUE THREAD

**Issue by constan1** [Plachtaa/VALL-E-X#8]
Great implementation. In the paper it says phoneme acoustic token pairs <S,A> are concatenated. So no decoding prefix is added? In the original valle they train on a prompt <x,C~,C<1 >  for both nar and ar model. 

In addition, are language ID's added elementwise to the both the ar model and nar model acoustic tokens during training?

**Comment by Plachtaa** [comment]
Both AR & NAR model receives prompt <S, A>, a prompt contains acoustic tokens sequences and its corresponding phoneme transcript.
Language ID's are added elementwise to deal with code-switch cases.

**Comment by Plachtaa** [comment]
1. This is because Microsoft was using LibriLight, which mainly comprises of speeches around 60 seconds, allowing them to sample one segment for AR model, and another segment as NAR model's prompt. But the dataset we use contains only short speeches below 30 seconds, so it is difficult to train with the same method as MS's;
2. Language embeddings are added only to phoneme tokens. This makes sense since the phonemes are IPAs.

**Comment by constan1** [comment]
I see. Choosing utterance of the same speaker isn't too difficult for labeled audio data.  Since each file is named by their speaker ID. Also,  Vallex says it uses utterances of 12 seconds max.

Also in the paper it says it they embed Language ID into dense vectors and add to acoustic token embeddings. But in your Implementation I see you added it to phonemes embeddings. Sorry I'm relatively new to this but just trying to understand the implementation as their paper is often ambigous in certain parts.

Thanks again.

**Comment by Plachtaa** [comment]
You are correct, choosing utterance of the same speaker isn't difficult for labeled audio data. If there is a chance I can start over again, I will follow your advice to use the same prompting method as MS's. I'm sorry for my mistake here.

As for where the language embeddings should go, MS indeed said the language embeddings are added to acoustic tokens, but this does not make any sense. I cannot think of any theory to explain why, and how they add the language embedding in this way. Hence, I merely assume it is a writing mistake. 
Anyway, the reproduced model seems to functioning well and produces similar result in cross-lingual speech synthesis just like MS's original demo.
However, I still keep the curiosity of hearing from other people about how they explain this. If you have any brilliant understanding, I'm glad to listen to your explanation.

**Comment by constan1** [comment]
Thanks, im curious to see other implementations to broaden my understanding. 

So does the ar model use another utterance of the same speaker or only pairs <Ss,As:1> and <St,At:1> where these are source and target languages, respectively?  Here I figure the acoustic tokens in this case are the first layer A:1 for the ar model? The original valle uses the same utterance speaker for both ar and nar models, vallex appears  to use another utterance only for nar model?

**Comment by Plachtaa** [comment]
Sorry I'm not able to cover all the implementation details by typing. For more details, I'll suggest you find the answers from my codes and VALL-E X's paper, they should be more clear and straightforward.

**Comment by Plachtaa** [comment]
Close this issue due to long term no activity of issue creator.,
